samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -208,25 +179,30 @@ class KafkaSystemProducer(systemName: String,
       metrics.flushes.inc
 
       val sourceData = sources.get(source)
-      //if latestFuture is null, it probably means that there has been no calls to "send" messages
-      //Hence, nothing to do in flush
-      if(sourceData.latestFuture != null) {
-        while(!sourceData.latestFuture.isDone && sourceData.exceptionInCallback.get() == null) {
-          try {
-            sourceData.latestFuture.get()
-          } catch {
-            case t: Throwable => error(t.getMessage, t)
-          }
-        }
+      val currentProducer = producer
 
-        //if there is an exception thrown from the previous callbacks just before flush, we have to fail the container
-        if (sourceData.exceptionInCallback.get() != null) {
-          metrics.flushFailed.inc
-          throw sourceData.exceptionInCallback.get()
+      if (currentProducer == null) {
+        metrics.flushFailed.inc
+        val exception = sourceData.exceptionInCallback.getAndSet(null)
+        // Throw the original exception if possible. If not, throw exception for null producer.
+        if (exception != null) {
+          throw new SystemProducerException("Flush failed. A batch of messages were unsent!", exception)
         } else {
-          trace("Flushed %s." format (source))
+          throw new SystemProducerException("Kafka system producer is not available.")
         }
       }
+
+      currentProducer.flush()
```

Comment:
It's possible that we have a saved exception and a working new Producer here. We can probably pre-check for exception and throw before trying to flush the producer to avoid out-of-order delivery during flush. Doesn't prevent all OOO delivery (see comment above), but won't hurt. It does look awkward, but refactoring might help.